homeassistant/components/dio_chacon/__init__.py

```diff
@@ -0,0 +1,98 @@
+"""The dio_chacon integration."""
+
+import logging
+from typing import Any
+
+from dio_chacon_wifi_api import DIOChaconAPIClient
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_PASSWORD,
+    CONF_UNIQUE_ID,
+    CONF_USERNAME,
+    EVENT_HOMEASSISTANT_STOP,
+    Platform,
+)
+from homeassistant.core import Event, HomeAssistant, ServiceCall
+
+from .const import (
+    DOMAIN,
+    EVENT_DIO_CHACON_DEVICE_STATE_CHANGED,
+    EVENT_DIO_CHACON_DEVICE_STATE_RELOAD,
+)
+
+SERVICE_RELOAD_STATE = "reload_state"
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.COVER]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up dio_chacon from a config entry."""
+
+    _LOGGER.debug("Start of async_setup_entry for dio_chacon integration")
+
+    hass.data.setdefault(DOMAIN, {})
+
+    config = entry.data
+
+    def callback_device_state(data: Any) -> None:
+        """Receive callback for device state notification pushed from the server."""
+
+        hass.bus.fire(EVENT_DIO_CHACON_DEVICE_STATE_CHANGED, data)
```

Comment:
The design of the API is not REST based but it is based on a permanent connection via Websocket (established just after login).
I tried to explain it here in the design principle : https://github.com/cnico/dio-chacon-wifi-api
This callback function is asynchronously called from the websocket server side (for example when a cover is manually moved) and it helps keeping the state in home assistant coherent.
When homeassistant starts, it will logs in and establish the websocket connection permanently. This callback is linked to received server side events whenever they arrive.
It was inspired by the plex integration implementation and its plex_websocket_callback.
The EVENT_DIO_CHACON_DEVICE_STATE_CHANGED is send to all platforms (this pr does not contain the switch one) and the event is effectively used to update the cibled entity.
I am not expert enough but I don't find how to use a coordinator in this situation.

Reply:
You can use a coordinator as the data owner, where you let the websocket update the data of the coordinator, which efficiently updates the other entities

Reply:
Implemented it with a coordinator : it works correctly with my local tests.